### ISSUE TITLE
Index out of range for attribute `Obj`

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaOpt/Gurobi.jl#195]
Hello all, I'm running on master as whatever performance issues exist for MOI are prohibitive for me on the tagged version.  I'm getting the following error on `optimize!`
```
ERROR: Gurobi.GurobiError(10006, "Index out of range for attribute 'Obj'")
Stacktrace:
 [1] set_dblattrarray!(::Gurobi.Model, ::String, ::Int64, ::Int64, ::Array{Float64,1}) at /home/savastio/.julia/dev/Gurobi/src/grb_attrs.jl:307
 [2] set_linear_objective!(::Gurobi.Optimizer, ::Array{Int64,1}, ::Array{Float64,1}) at /home/savastio/.julia/dev/Gurobi/src/MOI_wrapper.jl:366
 [3] unsafe_set!(::Gurobi.Optimizer, ::MathOptInterface.ObjectiveFunction{MathOptInterface.ScalarAffineFunction{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/objective.jl:77
 [4] set at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/objective.jl:61 [inlined]
 [5] set(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.ObjectiveFunction{MathOptInterface.ScalarAffineFunction{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:195
 [6] _pass_attributes(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool, ::MathOptInterface.Utilities.IndexMap, ::Array{MathOptInterface.AbstractModelAttribute,1}, ::Tuple{}, ::Tuple{}, ::Tuple{}, ::typeof(MathOptInterface.set)) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:141
 [7] pass_attributes at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:105 [inlined]
 [8] pass_attributes at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:104 [inlined]
 [9] default_copy_to(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:195
 [10] #automatic_copy_to#61 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:15 [inlined]
 [11] #automatic_copy_to at ./none:0 [inlined]
 [12] #copy_to#1 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:91 [inlined]
 [13] (::getfield(MathOptInterface, Symbol("#kw##copy_to")))(::NamedTuple{(:copy_names,),Tuple{Bool}}, ::typeof(MathOptInterface.copy_to), ::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}) at ./none:0
 [14] attach_optimizer(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:125
 [15] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:161
 [16] #optimize!#77(::Bool, ::Bool, ::Function, ::Model, ::Nothing) at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:132
 [17] optimize! at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:105 [inlined] (repeats 2 times)
 [18] top-level scope at none:0
```
Is this obvious or do I need more digging?

**Comment by odow** [comment]
Oops. We just need tag. I'll hold off until this is resolved though.

Do you have a reproducible example?

**Comment by ExpandingMan** [comment]
Unfortunately my example is very convoluted, so it might be hard to track it down.  Any ideas where I should start looking?  (perhaps in `set_linear_objective`? that seems straightforward enough)

Unfortunately I'm rather out of the loop on MOI, I'm going to start reading up again this evening, really hoping we'll be able to get this fixed in the next couple of days.

**Comment by odow** [comment]
Try: https://github.com/JuliaOpt/Gurobi.jl/pull/196

In order to get some performance, we have to cleverly call `update`. Hopefully I just missed one.

**Comment by ExpandingMan** [comment]
#196 prevents it from crashing, but it still seems to have extremely severe performance issues.  When I call to `optimize!` I get to "Server capacity available..." but then it seems to sit and spin forever.  Here is where it was when I interrupted it:
```
^CERROR: InterruptException:
Stacktrace:
 [1] objectid at ./reflection.jl:297 [inlined]
 [2] hash at ./hashing.jl:23 [inlined]
 [3] hash at ./hashing.jl:18 [inlined]
 [4] hashindex at ./dict.jl:169 [inlined]
 [5] ht_keyindex2!(::Dict{MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.ZeroOne},Tuple{MathOptInterface.VariableIndex,Float64,Float64}}, ::MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.ZeroOne}) at ./dict.jl:309
 [6] setindex!(::Dict{MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.ZeroOne},Tuple{MathOptInterface.VariableIndex,Float64,Float64}}, ::Tuple{MathOptInterface.VariableIndex,Float64,Float64}, ::MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.ZeroOne}) at ./dict.jl:382
 [7] add_constraint(::Gurobi.Optimizer, ::MathOptInterface.SingleVariable, ::MathOptInterface.ZeroOne) at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/constraints/singlevariable.jl:208
 [8] add_constraint(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.SingleVariable, ::MathOptInterface.ZeroOne) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:323
 [9] copyconstraints!(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool, ::MathOptInterface.Utilities.IndexMap, ::Type{MathOptInterface.SingleVariable}, ::Type{MathOptInterface.ZeroOne}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:158
 [10] default_copy_to(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:200
 [11] #automatic_copy_to#61 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:15 [inlined]
 [12] #automatic_copy_to at ./none:0 [inlined]
 [13] #copy_to#1 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:91 [inlined]
 [14] (::getfield(MathOptInterface, Symbol("#kw##copy_to")))(::NamedTuple{(:copy_names,),Tuple{Bool}}, ::typeof(MathOptInterface.copy_to), ::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}) at ./none:0
 [15] attach_optimizer(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:125
 [16] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:161
 [17] #optimize!#77(::Bool, ::Bool, ::Function, ::JuMP.Model, ::Nothing) at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:132
 [18] optimize! at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:105 [inlined] (repeats 2 times)
```

**Comment by odow** [comment]
Can you share an example? 

Also, can you try GLPK? They share a lot of code so that should help us narrow down where the performance problem is.

**Comment by ExpandingMan** [comment]
This is a pretty elaborate problem, creating an MWE is not straightforward, but hopefully I can get there with just shear size.  Here's my model object
```
Minimization problem with:
Variables: 1165510
Objective function type: GenericAffExpr{Float64,VariableRef}
`VariableRef`-in-`MathOptInterface.ZeroOne`: 131982 constraints
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 1165510 constraints
`VariableRef`-in-`MathOptInterface.LessThan{Float64}`: 850548 constraints
`GenericAffExpr{Float64,VariableRef}`-in-`MathOptInterface.LessThan{Float64}`: 1034891 constraints
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Gurobi
```
I'll try to post an MWE soon, then we can test that with GLPK.

**Comment by ExpandingMan** [comment]
By the way, the way I create the problem I have a lot of spurious
```
0 ≤ 1.0
0 ≤ 0.0
```
constraints.  Obviously this happens easily when you do some sparse matrix operations to formulate your problem.

In v0.18 they were elided properly and never caused any issues.  I do see them appearing in the model constraint counts above.  Is there any possibility that these are not being elided correctly?

**Comment by ExpandingMan** [comment]
The following seems to be an ok-ish MWE
```julia
using JuMP, Gurobi, Random

Random.seed!(999)

const MAX_TERMS = 100
const Nvars = 5*10^5 

function constraint_mat()
    M = zeros(Int, 3, Nvars)
    for i ∈ 1:size(M,1)
        # this is to get lots of spurious constraints
        if rand() < 0.5
            for j ∈ rand(1:Nvars, rand(1:MAX_TERMS))
                M[i, j] = 1
            end    
        end
    end
    M
end

m = Model(with_optimizer(Gurobi.Optimizer))

@variable(m, x[1:Nvars], Bin)

@objective(m, Min, -sum(x))

M = constraint_mat()
@constraint(m, M*x .≤ 1)

optimize!(m)  # expect hang here
```

**Comment by odow** [comment]
I'm working on what I think is a fix. 

To confirm, please try a smaller model, then look at the Gurobi log and see if it complains about update being called too frequently.

**Comment by odow** [comment]
Please try #196

**Comment by ExpandingMan** [comment]
I just tried it with
```julia
@variable(m, 0 ≤ x[1:Nvars] ≤ 5, Int)
```
and it seemed to work fine, so I think you're on to something.  Let me try it with your updated PR.

**Comment by ExpandingMan** [comment]
Seem to be making progress!  The MWE seems to work on your PR, but when I go back to my real problem I get
```ERROR: Cannot set bounds because variable is of type: BINARY.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] add_constraint(::Gurobi.Optimizer, ::MathOptInterface.SingleVariable, ::MathOptInterface.GreaterThan{Float64}) at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/constraints/singlevariable.jl:112
 [3] add_constraint(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.SingleVariable, ::MathOptInterface.GreaterThan{Float64}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:323
 [4] copyconstraints!(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool, ::MathOptInterface.Utilities.IndexMap, ::Type{MathOptInterface.SingleVariable}, ::Type{MathOptInterface.GreaterThan{Float64}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:158
 [5] default_copy_to(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:200
 [6] #automatic_copy_to#61 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:15 [inlined]
 [7] #automatic_copy_to at ./none:0 [inlined]
 [8] #copy_to#1 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:91 [inlined]
 [9] (::getfield(MathOptInterface, Symbol("#kw##copy_to")))(::NamedTuple{(:copy_names,),Tuple{Bool}}, ::typeof(MathOptInterface.copy_to), ::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}) at ./none:0
 [10] attach_optimizer(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:125
 [11] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:161
 [12] #optimize!#77(::Bool, ::Bool, ::Function, ::Model, ::Nothing) at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:132
 [13] optimize! at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:105 [inlined] (repeats 2 times)
 [14] top-level scope at util.jl:156
```

**Comment by odow** [comment]
What's the error?

**Comment by ExpandingMan** [comment]
Oh sorry, copy-paste catastrophe.  It's
```
ERROR: Cannot set bounds because variable is of type: BINARY.
```

**Comment by odow** [comment]
It looks like we decided in https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/31 that you shouldn't be able to add bounds to a binary variable. We should probably relax that... But what are the bounds on your binary variables?

**Comment by ExpandingMan** [comment]
Yes, that really should be relaxed, my use case is likely a common one: some sparse matrices get multiplied together, the result of which is a matrix with lots of simple constraints such as
```julia
x + y + z ≤ 1
```
However, it seems likely that, just by chance, there is a small but non-zero subset of these where one winds up with
```julia
x ≤ 1
```

In general I would say that it is really important to be able to simply elide strange or trivial looking constraints in case they get generated as the result of simple linear algebra operations.

**Comment by odow** [comment]
> In general I would say that it is really important to be able to simply elide strange or trivial looking constraints in case they get generated as the result of simple linear algebra operations.

I don't think this is in JuMP's mandate. How are we to distinguish what is trivial? Maybe you want to add a dummy constraint now, but modify the coefficients later (e.g., column generation).

Here's the PR in LQOI: https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/89

**Comment by ExpandingMan** [comment]
> I don't think this is in JuMP's mandate. How are we to distinguish what is trivial? Maybe you want to add a dummy constraint now, but modify the coefficients later (e.g., column generation).

Fair enough.  I guess what I ought to say is that if JuMP is so restrictive that it errors out on many cases which are likely to result form some sort of vectorized or automated generation of constraints, it will quickly become useless.  As such, perhaps it is best to be fairly liberal with these things and let users suffer with some inefficiency if they are causing it in the way they generate constraints.

**Comment by ExpandingMan** [comment]
I've tracked down where I'm adding the constraints to the binary values and disabled them.  Unfortunately, we seem to be right back at (the same?) performance problem, it always is stuck here
```
^CERROR: InterruptException:
Stacktrace:
 [1] objectid at ./reflection.jl:297 [inlined]
 [2] hash at ./hashing.jl:23 [inlined]
 [3] hash at ./hashing.jl:18 [inlined]
 [4] hashindex at ./dict.jl:169 [inlined]
 [5] ht_keyindex2!(::Dict{MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}},Int64}, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}) at ./dict.jl:309
 [6] setindex!(::Dict{MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}},Int64}, ::Int64, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}) at ./dict.jl:382
 [7] add_constraint(::Gurobi.Optimizer, ::MathOptInterface.ScalarAffineFunction{Float64}, ::MathOptInterface.LessThan{Float64}) at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/constraints/scalaraffine.jl:19
 [8] add_constraint(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.ScalarAffineFunction{Float64}, ::MathOptInterface.LessThan{Float64}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:323
 [9] copyconstraints!(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool, ::MathOptInterface.Utilities.IndexMap, ::Type{MathOptInterface.ScalarAffineFunction{Float64}}, ::Type{MathOptInterface.LessThan{Float64}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:158
 [10] default_copy_to(::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}, ::Bool) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:200
 [11] #automatic_copy_to#61 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/copy.jl:15 [inlined]
 [12] #automatic_copy_to at ./none:0 [inlined]
 [13] #copy_to#1 at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Bridges/bridgeoptimizer.jl:91 [inlined]
 [14] (::getfield(MathOptInterface, Symbol("#kw##copy_to")))(::NamedTuple{(:copy_names,),Tuple{Bool}}, ::typeof(MathOptInterface.copy_to), ::MathOptInterface.Bridges.LazyBridgeOptimizer{Gurobi.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}) at ./none:0
 [15] attach_optimizer(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:125
 [16] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /home/savastio/.julia/packages/MathOptInterface/QVlW6/src/Utilities/cachingoptimizer.jl:161
 [17] #optimize!#77(::Bool, ::Bool, ::Function, ::Model, ::Nothing) at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:132
 [18] optimize! at /home/savastio/.julia/packages/JuMP/jnmGG/src/optimizer_interface.jl:105 [inlined] (repeats 2 times)
```

Unfortunately my original MWE can no longer reproduce this issue.  I'll figure out an updated one tomorrow.

**Comment by mlubin** [comment]
@ExpandingMan You should try also JuMP's direct mode (i.e., `direct_model`). That's intended to be as fast as using Gurobi's C++ API directly with no layers in between.

**Comment by odow** [comment]
> That's intended to be as fast as using Gurobi's C++ API directly with no layers in between.

It has the same problem because we don't expose the `Gurobi.update` to the user. Maybe we should have a `MOI.set(model, Gurobi.UpdateModel())` and a `Gurobi.Optimizer(;automatic_update = true)` option.

**Comment by odow** [comment]
This re-write of the `SingleVariable` constraints in LQOI should help: https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/88

**Comment by ExpandingMan** [comment]
I'm back to getting the error
```
ERROR: LoadError: Gurobi.GurobiError(10006, "Index out of range for attribute 'LB'")
Stacktrace:
 [1] set_dblattrelement!(::Gurobi.Model, ::String, ::Int64, ::Float64) at /home/savastio/.julia/dev/Gurobi/src/grb_attrs.jl:260
 [2] add_constraint(::Gurobi.Optimizer, ::MathOptInterface.SingleVariable, ::MathOptInterface.GreaterThan{Float64}) at /home/savastio/.julia/dev/Gurobi/src/MOI_wrapper.jl:180
 [3] set_lower_bound(::VariableRef, ::Int64) at /home/savastio/.julia/packages/JuMP/jnmGG/src/variables.jl:358
 [4] add_variable(::Model, ::ScalarVariable{Int64,Float64,Float64,Float64}, ::String) at /home/savastio/.julia/packages/JuMP/jnmGG/src/variables.jl:745
 [5] top-level scope at /home/savastio/.julia/packages/JuMP/jnmGG/src/macros.jl:177
 [6] include at ./boot.jl:326 [inlined]
 [7] include_relative(::Module, ::String) at ./loading.jl:1038
 [8] include(::Module, ::String) at ./sysimg.jl:29
 [9] include(::String) at ./client.jl:403
 [10] top-level scope at none:0
in expression starting at /home/savastio/src/jump_test.jl:10
```
when using `direct_model`.  I take it the same update line needs to be added elsewhere?

**Comment by odow** [comment]
What version of Gurobi are you using?

**Comment by odow** [comment]
Also, can you provide reproducible examples for all of these errors? I can't reproduce this. The following works without error:
```julia
model = JuMP.direct_model(Gurobi.Optimizer())
@variable(model, x)
set_lower_bound(x, 1)
@variable(model, y >= 2)
optimize!(model)
```

**Comment by ExpandingMan** [comment]
This gives the above error for me on #196 
```julia
using JuMP, Gurobi

const N = 10

# m = Model(with_optimizer(Gurobi.Optimizer, MIPGap=0.05, MIPFocus=1))
m = direct_model(Gurobi.Optimizer(MIPGap=0.05, MIPFocus=1))

@variable(m, x[1:N,1:N] ≥ 0, Int)
M = 2ones(Int, N, N)

@objective(m, Min, -sum(x))

@constraint(m, UpperBound, x .≤ M)
```
Unfortunately we're still on Gurobi 6.52.  I'd really like to upgrade it, but I'm not sure when we'll be able to do that.

**Comment by odow** [comment]
Bingo. So the problem is that sometime between Gurobi 6.5.2 and Gurobi 8.0, they changed how the lazy `update` worked...

Looking at 16) in http://www.gurobi.com/support/faqs, it seems that this was Gurobi 7.0. Please try with 
`Gurobi.Optimizer(UpdateMode = 1)`.

**Comment by ExpandingMan** [comment]
In that case I get
```
ERROR: LoadError: Gurobi.GurobiError(10022, "Problem updating model")
Stacktrace:
 [1] update_model! at /home/savastio/.julia/dev/Gurobi/src/grb_model.jl:102 [inlined]
 [2] _update_if_necessary at /home/savastio/.julia/dev/Gurobi/src/MOI_wrapper.jl:110 [inlined]
 [3] set_linear_objective!(::Gurobi.Optimizer, ::Array{Int64,1}, ::Array{Float64,1}) at /home/savastio/.julia/dev/Gurobi/src/MOI_wrapper.jl:366
 [4] unsafe_set!(::Gurobi.Optimizer, ::MathOptInterface.ObjectiveFunction{MathOptInterface.ScalarAffineFunction{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/objective.jl:77
 [5] set(::Gurobi.Optimizer, ::MathOptInterface.ObjectiveFunction{MathOptInterface.ScalarAffineFunction{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/savastio/.julia/packages/LinQuadOptInterface/ZMx9f/src/objective.jl:61
 [6] set(::Model, ::MathOptInterface.ObjectiveFunction{MathOptInterface.ScalarAffineFunction{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/savastio/.julia/packages/JuMP/jnmGG/src/JuMP.jl:621
 [7] set_objective_function(::Model, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/savastio/.julia/packages/JuMP/jnmGG/src/objective.jl:66
 [8] set_objective_function(::Model, ::GenericAffExpr{Float64,VariableRef}) at /home/savastio/.julia/packages/JuMP/jnmGG/src/objective.jl:77
 [9] set_objective(::Model, ::MathOptInterface.OptimizationSense, ::GenericAffExpr{Float64,VariableRef}) at /home/savastio/.julia/packages/JuMP/jnmGG/src/objective.jl:83
 [10] top-level scope at /home/savastio/.julia/packages/JuMP/jnmGG/src/macros.jl:977
 [11] include at ./boot.jl:326 [inlined]
 [12] include_relative(::Module, ::String) at ./loading.jl:1038
 [13] include(::Module, ::String) at ./sysimg.jl:29
 [14] include(::String) at ./client.jl:403
 [15] top-level scope at none:0
in expression starting at /home/savastio/src/jump_test.jl:11
```
where I used the same code as above except with `Gurobi.Optimizer(UpdateMode=1)`

**Comment by odow** [comment]
Documentation for 6.5 is no longer available online. However, on 8.0, 10022 is a network error:
http://www.gurobi.com/documentation/8.0/refman/error_codes.html

IMO, I think we should remove support for Gurobi version < 7. If they're no longer supported by Gurobi, I don't see why we should. (Unless someone wants to maintain and test on those versions.)
https://github.com/JuliaOpt/Gurobi.jl/blob/98df256c61c9928737d1bd14215ac677311e8723/deps/build.jl#L25-L30

**Comment by ExpandingMan** [comment]
> IMO, I think we should remove support for Gurobi version < 7. If they're no longer supported by Gurobi, I don't see why we should. (Unless someone wants to maintain and test on those versions.)

Fair enough.  It's pretty clear we are going to have to update.

**Comment by ExpandingMan** [comment]
Ok, the good news is that we're now on Gurobi 8.1 and we should never again be in a position where it's really hard for us to update Gurobi on our dev environment.  The bad news is, I'm having the same performance issue, even with `direct_model`, except that now of course it occurs when I'm actually declaring the constraints.  For the record, on JuMP 0.18 it has no problem at all and is quite fast.

I'm going to try to figure out a proper MWE tomorrow.  Since the issue this thread was opened for was solved by #196 (or will be when it's merged), should we close this issue and I'll open one tomorrow with a MWE for the performance issue?

In the meantime, just FYI it still appears to be getting stuck in the same place:
```
^CERROR: LoadError: InterruptException:
Stacktrace:
 [1] objectid at ./reflection.jl:297 [inlined]
 [2] hash at ./hashing.jl:23 [inlined]
 [3] hash at ./hashing.jl:18 [inlined]
 [4] hashindex at ./dict.jl:169 [inlined]
 [5] ht_keyindex2!(::Dict{MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}},Int64}, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}) at ./dict.jl:309
 [6] setindex!(::Dict{MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}},Int64}, ::Int64, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}) at ./dict.jl:382
 [7] add_constraint(::Gurobi.Optimizer, ::MathOptInterface.ScalarAffineFunction{Float64}, ::MathOptInterface.LessThan{Float64}) at /home/expandingman/.julia/packages/LinQuadOptInterface/ZMx9f/src/constraints/scalaraffine.jl:19
 [8] moi_add_constraint(::Gurobi.Optimizer, ::MathOptInterface.ScalarAffineFunction{Float64}, ::MathOptInterface.LessThan{Float64}) at /home/expandingman/.julia/packages/JuMP/jnmGG/src/constraints.jl:373
 [9] add_constraint(::JuMP.Model, ::JuMP.ScalarConstraint{JuMP.GenericAffExpr{Float64,JuMP.VariableRef},MathOptInterface.LessThan{Float64}}, ::String) at /home/expandingman/.julia/packages/JuMP/jnmGG/src/constraints.jl:385
 [10] _broadcast_getindex at /home/expandingman/.julia/packages/JuMP/jnmGG/src/constraints.jl:384 [inlined]
 [11] getindex at ./broadcast.jl:511 [inlined]
 [12] copyto_nonleaf!(::Array{JuMP.ConstraintRef{JuMP.Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}},JuMP.ScalarShape},2}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{2},Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(JuMP.add_constraint),Tuple{Base.RefValue{JuMP.Model},Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{2},Nothing,typeof(JuMP.build_constraint),Tuple{Base.RefValue{getfield(JuMP, Symbol("#_error#55")){Symbol}},Base.Broadcast.Extruded{Array{JuMP.GenericAffExpr{Float64,JuMP.VariableRef},2},Tuple{Bool,Bool},Tuple{Int64,Int64}},Base.RefValue{MathOptInterface.LessThan{Float64}}}}}}, ::CartesianIndices{2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}}, ::CartesianIndex{2}, ::Int64) at ./broadcast.jl:928
```

**Comment by odow** [comment]
I think I know the problem.
